code/modules/events/wormholes.dm

```diff
@@ -0,0 +1,52 @@
+/datum/event/wormholes
+	announceWhen = 10
+	endWhen = 60
+
+	var/list/pick_turfs = list()
+	var/list/wormholes = list()
+	var/shift_frequency = 3
+	var/number_of_wormholes = 400
+
+/datum/event/wormholes/setup()
+	announceWhen = rand(0, 20)
+	endWhen = rand(40, 80)
+
+/datum/event/wormholes/start()
+	var/list/areas = area_repository.get_areas_by_z_level()
+	for(var/i in areas)
+		var/area/A = areas[i]
+		for(var/turf/simulated/floor/T in A)
+			if(isAdminLevel(T.z))
+				continue
+			if(turf_contains_dense_objects(T))
+				continue
+			pick_turfs += T
+
+	for(var/i in 1 to number_of_wormholes)
```

Comment:
`to min(length(pick_turfs), number_of_wormholes)` in case you don't have 400 available turfs.